infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/msgHandlers/LogReplicationServer.java

```diff
@@ -0,0 +1,388 @@
+ * Replication entry messages, which represent parts of a Snapshot (full) sync or a Log Entry (delta) sync and also handles negotiation messages,
+ */
+@Slf4j
+public class LogReplicationServer extends LogReplicationAbstractServer {
+
+    // Unique and immutable identifier of a server node (UUID)
+    // Note: serverContext.getLocalEndpoint() can return an IP or FQDN, which is mutable (for this we
+    // should have a unique way the identify a node in the  topology
+    private String localNodeId;
+
+    // Cluster Id of the local node.
+    private String localClusterId;
+
+    private final ExecutorService executor;
+
+    private static final String EXECUTOR_NAME_PREFIX = "LogReplicationServer-";
+
+    private Map<LogReplicationSession, LogReplicationSinkManager> sessionToSinkManagerMap = new HashMap<>();
+
+    private final AtomicBoolean isLeader = new AtomicBoolean(false);
+
+    @Getter
+    private SessionManager sessionManager;
+
+    private ServerContext serverContext;
+
+    private String localEndpoint;
+
+    /**
+     * RequestHandlerMethods for the LogReplication server
+     */
+    @Getter
+    private final ReplicationHandlerMethods handlerMethods = createHandlerMethods();
+
+    protected ReplicationHandlerMethods createHandlerMethods() {
+        return ReplicationHandlerMethods.generateHandler(MethodHandles.lookup(), this);
+    }
+
+    // Shama check if you really need ServerContext
+    public LogReplicationServer(@Nonnull ServerContext context, @Nonnull SessionManager sessionManager,
+                                String localEndpoint) {
+        this.serverContext = context;
+        this.localEndpoint = localEndpoint;
+        this.localNodeId = sessionManager.getTopology().getLocalNodeDescriptor().getNodeId();
+        this.localClusterId = sessionManager.getTopology().getLocalClusterDescriptor().getClusterId();
+        this.sessionManager = sessionManager;
+        createSinkManagers();
+        this.executor = context.getExecutorService(1, EXECUTOR_NAME_PREFIX);
+    }
+
+    @VisibleForTesting
+    public LogReplicationServer(@Nonnull ServerContext context, LogReplicationSinkManager sinkManager,
+        String localNodeId, String localClusterId, SessionManager sessionManager) {
+        this.serverContext = context;
+        this.localEndpoint = null;
+        sessionToSinkManagerMap.put(sinkManager.getSession(), sinkManager);
+        this.localNodeId = localNodeId;
+        this.localClusterId = localClusterId;
+        this.sessionManager = sessionManager;
+        this.executor = context.getExecutorService(1, EXECUTOR_NAME_PREFIX);
+    }
+
+     private void createSinkManagers() {
+        for (LogReplicationSession session : sessionManager.getIncomingSessions()) {
+            createSinkManager(session);
+        }
+    }
+
+    private LogReplicationSinkManager createSinkManager(LogReplicationSession session) {
```

Comment:
SinkManager not being returned.  The return type can be void

Reply:
The sinkManager is being returned for this method. https://github.com/CorfuDB/CorfuDB/pull/3560/files#diff-dc7bfb14445507a6bd90cee4558b5e576bbe2200cc33cfe1b5f183f66222beb4R116